[Merged by Bors] - 3315 Genesis address format - address size and type change

abergasov opened the pull request:
## Motivation
<!-- Please mention the issue fixed by this PR or detailed motivation -->
Closes #3315
<!-- `Closes #XXXX, closes #XXXX, ...` links mentioned issues to this PR and automatically closes them when this it's merged -->

## Changes
* address moved to separate pkg. probably in future will be as independent library
* address still array of bytes ([24]byte), but implement `bech32` algorithm. Additional first byte reserved for contain HumanReadablePart of address (now contain info about networkID - test|main net), so result is [25]byte
* all api endpoints will be changed to use string instead []byte (related pr https://github.com/spacemeshos/api/pull/173)

## Test Plan
unit tests

## TODO
<!-- This section should be removed when all items are complete -->
- [ ] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [ ] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [ ] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

abergasov committed:
address to separate pkg with bech32 support, hrp and new length

dshulyak commented:
could you please use aliases for such refactoring?

```go
package types

type Address = address.Address
```

they were created specifically to make such changes reviewable

abergasov committed:
constants replace

abergasov committed:
use alias for import fix

dshulyak reviewed (changes requested):
hrp part wont be stored on the mesh, even if it is 1 byte. it has other implications beside space

moreover:
- there will be no network ids. they will be replaced by hashes instead. so the idea to compress hrp wont work, but it is not even important
- there is no reason to hardcode every single known testnet name in advance. there are solutions that avoid that

abergasov committed:
Merge branch 'develop' into 3315-genesis-address-format

# Conflicts:
#	common/types/block_test.go
#	sql/transactions/transactions_test.go
#	systest/testcontext/context.go
#	txs/cache_test.go
#	txs/conservative_state_test.go
#	txs/mocks/mocks.go
#	txs/store_test.go

abergasov committed:
keep network info as part of address pkg config

abergasov committed:
Merge branch 'develop' into 3315-genesis-address-format

abergasov committed:
back to common/types pkg

additional tests and new test network prefix

drom redunant networkName

drop 0x hardcoded addresses

reserve 4 byte on top of address

rmv unused method

drop unessesarry logs in services, rename fromString

dshulyak commented on common/types/address/address_test.go:
```diff
@@ -0,0 +1,131 @@
+	return acc1Signer.PublicKey().Bytes()
+}
+
+//func intToAddress(t *testing.T, num uint32) *address.Address {
```
rm

dshulyak commented on api/grpcserver/mesh_service.go:
```diff
@@ -178,7 +179,11 @@ func (s MeshService) AccountMeshDataQuery(ctx context.Context, in *pb.AccountMes
-	addr := types.BytesToAddress(in.Filter.AccountId.Address)
+	addr, err := address.BytesToAddress(in.Filter.AccountId.Address)
+	if err != nil {
+		log.Error(fmt.Sprintf("failed to parse Filter.AccountId address: %v", err))
```
please stick to logs with fields

also:
- this is a warning, not the error
- it might be already logged by grpc middleware. please remove such logs if this is the case

dshulyak commented on common/types/address/address.go:
```diff
@@ -0,0 +1,156 @@
+type Address [AddressLength]byte // contains slice 8 bytes unsigned integers
+
+// NewAddress returns a new Address from a byte slice.
+func NewAddress(src string) (Address, error) {
```
FromString

lrettig replied:
NewAddressFromString?

dshulyak replied:
address is a part of the module, so what looks better when you use it?

address.NewAddressFromString or address.FromString

ps: i think the 1st one is incredibly ugly)

dshulyak commented on common/types/address/config.go:
```diff
@@ -0,0 +1,26 @@
+
+// Config is the configuration of the address package.
+type Config struct {
+	NetworkName string `mapstructure:"network-name"`
```
i think only hrp should be required to configure. why do we need NetworkName?

abergasov replied:
we do not need it. Probably we will later stream it via api for better ui usability?

dshulyak replied:
i don't know why we would need to stream it. for now it just looks like totally unnecessary config option that will have be changed for every network. lets remove it

dshulyak commented on cmd/node/node.go:
```diff
@@ -597,8 +598,12 @@ func (app *App) initServices(ctx context.Context,
 	if app.Config.SMESHING.Start {
+		coinbaseAddr, err = address.NewAddress(app.Config.SMESHING.CoinbaseAccount)
+		if err != nil {
+			app.log.Panic("failed to parse CoinbaseAccount address `%s`: %v", app.Config.SMESHING.CoinbaseAccount, err)
```
clearly can't panic in API request, return error to notify node administrator that node is miconfigured

abergasov replied:
this is not an api request, it node starting. On next lines, panic can happen if address check failing
```
if coinbaseAddr.Big().Uint64() == 0 {
	app.log.Panic("invalid coinbase account")
}
```

dshulyak replied:
sorry :) for some reason i thought that i saw same line in two places

dshulyak commented on api/config/genesis.go:
```diff
@@ -15,25 +19,39 @@ type GenesisConfig struct {
+		)
+		// this condition need in case address in config with specific hrp prefix, like smt1abc...
+		// in code we have constants which contain only bytes, without hrp prefix.
+		if strings.HasPrefix(addr, "0x") {
```
please remove this. either:

- use default hrp for this addresses
- check where this constants are used and remove them

abergasov replied:
in `address.BytesToAddress` will be used default hrp from `address.Config`

this consts are in `api/config/genesis.go`
they are used in `DefaultGenesisConfig` function 
```
type GenesisConfig struct {
	Accounts map[string]uint64 `mapstructure:"accounts"`
}

// Account1Address is the address from Account1Private.
const Account1Address = "0x0002abedccd3ffcbf46..."

// Account2Address is the address from Account2Address.
const Account2Address = "0xfeaf3d38b6ef43093..."

func DefaultGenesisConfig() *GenesisConfig {
	// NOTE(dshulyak) keys in default config are used in some tests
	g := GenesisConfig{}

	// we default to 10^5 SMH per account which is 10^17 smidge
	// each genesis account starts off with 10^17 smidge
	g.Accounts = map[string]uint64{
		Account1Address: 100000000000000000,
		Account2Address: 100000000000000000,
	}
	return &g
}
```

while able situation, when user can override with bech32 address like `sm1abc...`. 
So we should add limitation for config genesis address (like not use bech32 in address, but put `0x0123` string for genesis addr or other encoding style). Or just use such `if else`

dshulyak replied:
sorry i don't follow. please replace 0x addresses with addresses prefixed using hrp and remove this if condition. spacecraft and system tests have to be adjusted accordingly.

dshulyak commented on api/grpcserver/globalstate_service.go:
```diff
@@ -203,7 +212,11 @@ func (s GlobalStateService) AccountDataStream(in *pb.AccountDataStreamRequest, s
 		return status.Errorf(codes.InvalidArgument, "`Filter.AccountDataFlags` must set at least one bitfield")
 	}
-	addr := types.BytesToAddress(in.Filter.AccountId.Address)
+	addr, err := address.BytesToAddress(in.Filter.AccountId.Address)
```
it should be string with hrp

abergasov replied:
it will be changed, when PR  https://github.com/spacemeshos/api/pull/173 will be merged
i can add cast to string, which will be excess after api method change signature, or just wait until pr merge, that i'll replace to `FromString`

dshulyak commented on api/grpcserver/globalstate_service.go:
```diff
@@ -203,7 +212,11 @@ func (s GlobalStateService) AccountDataStream(in *pb.AccountDataStreamRequest, s
-	addr := types.BytesToAddress(in.Filter.AccountId.Address)
+	addr, err := address.BytesToAddress(in.Filter.AccountId.Address)
+	if err != nil {
+		log.Error(fmt.Sprintf("failed to parse in.Filter.AccountId.Address address: %v", err))
```
remove or downgrade and use fields

dshulyak commented on api/grpcserver/mesh_service.go:
```diff
@@ -437,7 +442,11 @@ func (s MeshService) AccountMeshDataStream(in *pb.AccountMeshDataStreamRequest,
-	addr := types.BytesToAddress(in.Filter.AccountId.Address)
+	addr, err := address.BytesToAddress(in.Filter.AccountId.Address)
+	if err != nil {
+		log.Error(fmt.Sprintf("failed to parse in.Filter.AccountId.Address address: %v", err))
```
same

dshulyak commented on common/types/address/address.go:
```diff
@@ -0,0 +1,156 @@
+func (a Address) Big() *big.Int { return new(big.Int).SetBytes(a[:]) }
+
+// Hex returns an EIP55-compliant hex string representation of the address.
+func (a Address) Hex() string {
```
should be removed, this is replaced with bech32 string

dshulyak commented on common/types/address/address.go:
```diff
@@ -0,0 +1,156 @@
+// Short returns the first 7 characters of the address hex representation (incl. "0x"), for logging purposes.
+func (a Address) Short() string {
+	hx := a.String()
+	return hx[:util.Min(7, len(hx))]
```
this needs to be fixed, since String will include hrp

abergasov replied:
i removed this method, as it used for 1 plase in logs. Just change to `String`

dshulyak commented on common/types/address/address.go:
```diff
@@ -0,0 +1,156 @@
+// GenerateAddress generates an address from a public key.
+func GenerateAddress(publicKey []byte) Address {
+	var addr Address
+	addr.setBytes(publicKey)
```
please double check with @lrettig if we need to leave 4 top bytes empty, he mentioned it somewhere in the github issues

lrettig replied:
correct, thanks @dshulyak

abergasov replied:
done, added bytes reservation + tests

dshulyak reviewed (changes requested):
why cosmos/bech32 and not one from btcd repo? are they different?

abergasov commented:
> why cosmos/bech32 and not one from btcd repo? are they different?

source code almost the same. small difference between const|wrappers, but in generally code the same. it can easy replaced by fixing `go.mod`

i chose cosmos cause it's approach - to get necessary functions and types, for crypto projects. in case btcd repo - we need to vendor other blockchein implementation, that's look little bit weird for me.

countvonzero reviewed (commented):
i am generally surprised by the number of files changed for this PR.

what is the advantage of having a separate pkg for address?
in another words, what is the disadvantage of maintaining the flat structure inside common/types?

abergasov commented:
> i am generally surprised by the number of files changed for this PR.
> 
> what is the advantage of having a separate pkg for address? in another words, what is the disadvantage of maintaining the flat structure inside common/types?

alias `type Address = address.Address` in common/type reduce total edited files =)

one of idea was to move address implementation to separate library, which will be included as dependency.
in this case  i moved to nested pkg. when we will publish as single library - just change imports

lrettig commented:
> one of idea was to move address implementation to separate library, which will be included as dependency

This is not a terrible idea - do we think other projects will have this dependency? (And would they want to use just an address module without relying on all of go-spacemesh?) E.g., sm-repl, dash/explorer, etc.

countvonzero commented:
> > one of idea was to move address implementation to separate library, which will be included as dependency
> 
> This is not a terrible idea - do we think other projects will have this dependency? (And would they want to use just an address module without relying on all of go-spacemesh?) E.g., sm-repl, dash/explorer, etc.

but why? what is the problem with having the impl inside common/types/address.go? the functionality you are providing in the new pkg look highly similar to the old ones. @abergasov

lrettig commented on common/types/address/address_test.go:
```diff
@@ -0,0 +1,133 @@
+			src:  "smt1fejq2x3d79ukpkw06t7h6lndjuwzxdnj59npghsywyusj",
+		},
+		{
+			name: "wrong length",
```
please test too long as well as too short

also, bad bech32 formats (missing hrp, invalid characters, etc.)

abergasov replied:
done
but actually it will just bech32 lib tests) all this cases will generate err on decode and it immediately returns

lrettig commented on common/types/address/config.go:
```diff
@@ -0,0 +1,23 @@
+// DefaultTestAddressConfig returns the default test configuration of the address package.
+func DefaultTestAddressConfig() *Config {
+	conf = &Config{
+		NetworkHRP: "smt",
```
"smt" is too similar to "sm", this is too error-prone. per https://github.com/spacemeshos/pm/issues/106#issuecomment-1170509194, let's use "stest" for the testnet.

abergasov commented:
> This is not a terrible idea - do we think other projects will have this dependency? (And would they want to use just an address module without relying on all of go-spacemesh?) E.g., sm-repl, dash/explorer, etc.

there is no strong arguments that we should move it to separate pkg, only this 2:
1. our possibly moving `address` in separate lib in future 
1. current situation in `common/types`, where we have 25 different files aka 25 different types (there are total 33 files with tests) and some of this files contains lot of code - this look like place where we can add more pkg separation and make it more strict

if you think that it better move back to `common/types` - ok, i'll do. arguments for separate pkg not very strong

countvonzero commented:
> > This is not a terrible idea - do we think other projects will have this dependency? (And would they want to use just an address module without relying on all of go-spacemesh?) E.g., sm-repl, dash/explorer, etc.
> 
> there is no strong arguments that we should move it to separate pkg, only this 2:
> 
> 1. our possibly moving `address` in separate lib in future

i think this is a future thing and should be separated from this discussion.

> 2. current situation in `common/types`, where we have 25 different files aka 25 different types (there are total 33 files with tests) and some of this files contains lot of code - this look like place where we can add more pkg separation and make it more strict

right. but changing only 1 makes it even more confusing. i'd support if you make a sweeping change later to change all of them. but the code inside each file (tho large) are mostly very simple code... i personally don't think it's worth the effort now.
 
> if you think that it better move back to `common/types` - ok, i'll do. arguments for separate pkg not very strong

that will be my preference. and worry about separate library when the use case arise.

abergasov committed:
Merge branch 'develop' into 3315-genesis-address-format

abergasov committed:
create instead copy, fix compare with zero, drop hardcoded accounts

dshulyak commented:
bors try

dshulyak commented on common/types/address.go:
```diff
@@ -1,84 +1,120 @@
+
 // Address represents the address of a spacemesh account with AddressLength length.
-type Address [AddressLength]byte
+type Address [AddressLength]byte // contains slice 8 bytes unsigned integers
```
> contains slice 8 bytes unsigned integers

what does it mean? where are the integers?

abergasov replied:
by default slice is uint8, value can be stored from 0-256 (base 2^8=256)
for bech32 need convert to 5-bit integers (2^5=32). it still uint8, but value up to 32
```
bech32.ConvertBits(addr[:], 8, 5, true) -> before encode to `sm1abc`
bech32.ConvertBits(addr[:], 5, 8, true) -> after encode from `sm1abc`
```

after this conversion slice length make bigger. so we keep as usually, uint8

dshulyak replied:
please remove this comment. it doesn't store uint64 (8 bytes unsigned inteegers) in any shape. what you provided above is a specific encoding of the bech32 format

dshulyak commented on common/types/address.go:
```diff
@@ -1,84 +1,120 @@
-	a.SetBytes(b)
-	return a
+// FromString returns a new Address from a given string like `sm1abc...`.
+func FromString(src string) (Address, error) {
```
it needs to be more specific now, since the module was removed. maybe StringToAddress if there is already BytesToAddress

abergasov replied:
`BytesToAddress` will be dropped, when we merge api changes
I'll rename to `StringToAddress`

dshulyak commented on common/types/address.go:
```diff
@@ -109,6 +136,17 @@ func (a *Address) DecodeScale(d *scale.Decoder) (int, error) {
+	copy(addr[AddressLength-len(publicKey):], publicKey)
+	for i := 0; i < AddressReservedSpace; i++ {
+		addr[i] = 0
+	}
```
copy(addr[4:], publicKey[len(publicKey)-24:])

dshulyak commented on common/types/address.go:
```diff
@@ -1,84 +1,120 @@
+
+	if conf.NetworkHRP != hrp {
+		return addr, fmt.Errorf("wrong network id: expected `%s`, got `%s`: %w", conf.NetworkHRP, hrp, ErrUnsupportedNetwork)
+	}
```
it needs a check that first 4 bytes are empty

dshulyak commented on config/presets/testnet.go:
```diff
@@ -34,11 +36,11 @@ func testnet() config.Config {
+			"stest1jt38hng8nmgywqc8vg83gffgfsajkxn9wjwr39qka8kuj": 100000000000000000,
+			"stest15e933zkyfvsgkxj0qxpx9fequcuhkzf2n2lzg2spvwvlf": 100000000000000000,
+			"stest13zwpcc8z0z7xcnu3jm9fxkudvzk70e6d5zvnerqjhu96v": 100000000000000000,
+			"stest1ty8kqh4w58s9qyytp7lc6ddvc84wvsy4rcvjtas3w5rkm": 100000000000000000,
```
this are the same addresses just re-encoded or from different private keys?

abergasov replied:
yes, i make ` util.FromHex(str)` and create address from it
as we keep algorithm how convert bytes to address (get last 20 bytes from byte slice)  and we fill first 4 bytes with zero - we can just regenerate this addresses

abergasov commented:
bors try

dshulyak commented:
what about system tests?

abergasov commented:
> what about system tests?

it fails in same place `cluster.Reuse` with ctx timeout. Probably node fails to start in case invalid genessis|coinbase accounts

abergasov committed:
systest addresses fix

abergasov committed:
Merge branch 'develop' into 3315-genesis-address-format

abergasov committed:
systest fix

abergasov commented:
bors try

abergasov committed:
Merge branch 'develop' into 3315-genesis-address-format

# Conflicts:
#	txs/conservative_state_test.go

abergasov committed:
api fix for support strings instead slice of bytes

abergasov commented:
bors try

abergasov commented:
bors try

abergasov commented:
bors merge

abergasov commented:
bors merge

lrettig commented on api/grpcserver/http_server.go:
```diff
@@ -6,7 +6,7 @@ import (
 	"sync"
 
-	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	runtime2 "github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
```
Why rename this?

lrettig commented on common/types/address.go:
```diff
@@ -1,84 +1,121 @@
+	// ErrUnsupportedNetwork is returned when a network is not supported.
+	ErrUnsupportedNetwork = errors.New("unsupported network")
+	// ErrDecodeBech32 is returned when an error occurs during decoding bech32.
+	ErrDecodeBech32 = errors.New("error decode to bech32")
```
```suggestion
	ErrDecodeBech32 = errors.New("error decoding bech32")
```

lrettig commented on common/types/address.go:
```diff
@@ -1,84 +1,121 @@
+	// for encoding bech32 uses slice of 5-bit unsigned integers. convert it back it 8-bit uints.
+	dataConverted, err := bech32.ConvertBits(data, 5, 8, true)
+	if err != nil {
+		return addr, fmt.Errorf("error convert bits to 8 bits: %w", err)
```
```suggestion
		return addr, fmt.Errorf("error converting bech32 bits: %w", err)
```

lrettig commented on common/types/address.go:
```diff
@@ -1,84 +1,121 @@
+
+	result, err := bech32.Encode(conf.NetworkHRP, dataConverted)
+	if err != nil {
+		log.Panic("error encode to bech32: ", err.Error())
```
```suggestion
		log.Panic("error encoding to bech32: ", err.Error())
```

lrettig commented on common/types/address.go:
```diff
@@ -1,84 +1,121 @@
-	return a.Hex()
+	dataConverted, err := bech32.ConvertBits(a[:], 8, 5, true)
+	if err != nil {
+		log.Panic("error convert bits to 8 bits: ", err.Error())
```
```suggestion
		log.Panic("error converting bech32 bits: ", err.Error())
```

lrettig commented on common/types/address_test.go:
```diff
@@ -0,0 +1,149 @@
+			src:  "stest1qqqqqqy0dd83jemjmfj3ghjndxm0ndh0z2rymaqyp0gu3",
+		},
+		{
+			name: "corect address, but no reserved space",
```
```suggestion
			name: "correct address, but no reserved space",
```

lrettig commented on common/types/address_test.go:
```diff
@@ -0,0 +1,149 @@
+			err:  types.ErrMissingReservedSpace,
+		},
+		{
+			name: "incorrect cahrset", // chars iob is not supported by bech32.
```
```suggestion
			name: "incorrect charset", // chars `iob` not supported by bech32.
```

lrettig commented on common/types/address_test.go:
```diff
@@ -0,0 +1,149 @@
+			err:  types.ErrDecodeBech32,
+		},
+		{
+			name: "to big length",
```
```suggestion
			name: "too long",
```

lrettig commented on common/types/address_test.go:
```diff
@@ -0,0 +1,149 @@
+			err:  types.ErrWrongAddressLength,
+		},
+		{
+			name: "to small length",
```
```suggestion
			name: "too short",
```

lrettig commented on common/types/address_test.go:
```diff
@@ -0,0 +1,149 @@
+			err:  types.ErrWrongAddressLength,
+		},
+		{
+			name: "wrong network",
```
```suggestion
			name: "wrong network (hrp)",
```

abergasov committed:
typo fix

abergasov commented:
bors try

dshulyak commented:
i thought that it was merged. is there any reason why it is still open?

abergasov committed:
Merge branch 'develop' into 3315-genesis-address-format

# Conflicts:
#	api/config/genesis.go
#	blocks/generator_test.go
#	go.mod
#	go.sum
#	tortoise/model/core.go

abergasov committed:
merge fix

abergasov commented:
bors try

abergasov commented:
bors merge
